x-pack/osquerybeat/ext/osquery-extension/pkg/amcache/state/state.go

```diff
@@ -0,0 +1,143 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+//go:build windows
+
+package state
+
+import (
+	"sync"
+	"time"
+
+	"github.com/elastic/beats/v7/x-pack/osquerybeat/ext/osquery-extension/pkg/amcache/registry"
+	"github.com/elastic/beats/v7/x-pack/osquerybeat/ext/osquery-extension/pkg/amcache/tables"
+	"github.com/elastic/beats/v7/x-pack/osquerybeat/ext/osquery-extension/pkg/filters"
+	"github.com/elastic/beats/v7/x-pack/osquerybeat/ext/osquery-extension/pkg/logger"
+)
+
+// Expiration duration for the global state.
+const defaultExpirationDuration = 3 * time.Minute
+
+// Default path to the Amcache hive.
+const defaultHivePath = "C:\\Windows\\AppCompat\\Programs\\Amcache.hve"
+
+// GlobalState is a shared state object that holds cached Amcache entries.
+// It is needed because loading and parsing the Amcache hive can be slow,
+// so we want to avoid doing it on every query. Instead, we load it once and
+// cache the results for a configurable duration.
+//
+// the amcache hive has multiple tables that can be queried, and when joins
+// are performed, osquery may call the generate function for each table multiple
+// times per query.  This can lead to significant performance issues if the
+// hive is reloaded each time.  By caching the results in a global state object,
+// we can avoid this overhead.
+//
+// The cache is not populated until the first query is made, to avoid
+// unnecessary work if the tables are not used.  Additionally, the cache
+// is refreshed at query time if it has expired. But will not be updated until
+// the next query, even if it is expired.
+//
+// TODO: Make sure that keeping this data in memory is not a problem for osquerybeat
+//
+//	in general
+
+type CachedTables map[tables.TableName][]tables.Entry
+
+type AmcacheGlobalState struct {
+	Cache              CachedTables
+	HivePath           string
+	ExpirationDuration time.Duration
+	Lock               sync.RWMutex
+	LastUpdated        time.Time
+}
+
+// NewAmcacheGlobalState creates a new AmcacheGlobalState instance with the default configuration.
+func newAmcacheGlobalState(hivePath string, expirationDuration time.Duration) *AmcacheGlobalState {
+	cachedTables := make(CachedTables)
+	for _, amcacheTable := range tables.AllAmcacheTables() {
+		cachedTables[amcacheTable.Name] = make([]tables.Entry, 0)
+	}
+	return &AmcacheGlobalState{HivePath: hivePath, ExpirationDuration: expirationDuration, Cache: cachedTables}
+}
+
+// Global variables for the gInstance and a mutex to protect it.
+var (
+	gInstance *AmcacheGlobalState = newAmcacheGlobalState(defaultHivePath, defaultExpirationDuration)
```

Comment:
The `g` prefix for globals that is more common in C/C++ codebases is not idiomatic in Go. You can make this `instance`. Also you don't need to declare the type, and can simplify it to `instance = newAmcache...`.